### PR TITLE
Add test covering #460 - making sure capturing request body does not erase the body

### PR DESCRIPTION
**Pull request by gregkalapos**
In #460 we had a bug which was that once the agent captures the request body, the body itself was not visible anymore in the MVC action method. This is fixed in https://github.com/elastic/apm-agent-dotnet/pull/539. 

This PR adds a test that reproduces and tests that scenario.

~The new test will fail until https://github.com/elastic/apm-agent-dotnet/pull/539 is not merged (atm. build is in-progress, I wait for it to finish, then I'll merge).~ Update: #539 is merged to `master`, I rebased this PR, all test should be green.

### CODE REVIEW

File: sample/SampleAspNetCoreApp/Controllers/HomeController.cs [hunk 1]
```diff
@@ -39,25 +40,25 @@ public Task<IActionResult> Index()
 		{
 			return SafeCaptureSpan<IActionResult>(GetCaptureControllerActionAsSpanFromQueryString(),
 				"Index_span_name", "Index_span_type", async () =>
-			{
```
Comment (gregkalapos) [inline sample/SampleAspNetCoreApp/Controllers/HomeController.cs]: Just code style... we are working on avoiding this. https://github.com/elastic/apm-agent-dotnet/pull/536

**Comment by gregkalapos** [comment]
There is 1 new failing test: `TestJsonBodyRetrievalOnRequestFailureInAspNetCore – Elastic.Apm.AspNetCore.Tests.AspNetCoreDiagnosticListenerTest` - investigating it.

File: test/Elastic.Apm.AspNetCore.Tests/Helper.cs [hunk 2]
```diff
@@ -27,8 +27,6 @@ public static class Helper
 
 						app.UseDeveloperExceptionPage();
 
-						app.UseHsts();
```
Comment (gregkalapos) [inline test/Elastic.Apm.AspNetCore.Tests/Helper.cs]: This is somewhat unrelated, but this HSTS is not needed, so I removed it.

**Comment by iquirino** [comment]
SampleAspNetCoreApp -> we can reproduce trying to add a new row on "InsertDb Sample".

The request.Body was initially typed as: Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream

After request.EnableRewind() it get the type changed to: Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream

You consume all the request.Body and then return a copy of the stream alredy used.
But this will have the position to the end and will not able to rewind.

You need to mantain the rewinded version.

I think a good test is:

Check if the Stream Position=0 and CanSeek=true

**Comment by gregkalapos** [comment]
> SampleAspNetCoreApp -> we can reproduce trying to add a new row on "InsertDb Sample".
> 
> The request.Body was initially typed as: Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream
> 
> After request.EnableRewind() it get the type changed to: Microsoft.AspNetCore.WebUtilities.FileBufferingReadStream
> 
> You consume all the request.Body and then return a copy of the stream alredy used.
> But this will have the position to the end and will not able to rewind.
> 
> You need to mantain the rewinded version.
> 
> I think a good test is:
> 
> Check if the Stream Position=0 and CanSeek=true

I agree with your analysis, but I'm not sure I get what you suggest here. What I'm saying is that our original test did not repro this problem (probably due to the fact that we did not really have real HTTP requests going on), but the test in this PR does. The test we added here failed before I merged your fix.

>  Check if the Stream Position=0 and CanSeek=true

Yeah, that'd make sense, but given the current test also repros it I think it's fine. But feel free to submit a PR, I'm happy to take such a test, but I think the coverage with this new test is already good.

**Comment by gregkalapos** [comment]
> Test itself LGTM:
> 
> https://github.com/elastic/apm-agent-dotnet/blob/1247981652264355d5a900fa9476f9503661e186/src/Elastic.Apm.AspNetCore/Extensions/RequestExtentions.cs#L33-L39
> 
> The truncate here seems less then optimal.
> 
> Is capturing a body something worth putting behind a configuration toggle? `EnableRewind()` does things like spooling to disk.

By default this feature is disabled - so it's already behind a configuration. Yeah, I'm also not a big fan of this, but I think given that it's not on by default I think it's fine... maybe we can research better ways to do this later, e.g. `HttpContext.Request.BodyReader` in ASP.NET Core 3.0.

**Pull request merged**